Cross matches

javierggt opened the pull request:
This PR is a refactor of the cross-match methods, partly to address requests in issue  #15 and partly to fix a few things.

The things that changed upon @taldcroft request:
- Cross_match.cross_match changed so one can customize the query.
- Docstrings much improved.
- added methods to exclude regions/targets, and optionally exclude regions in cross_match

Other cleanup:
- Removed older functions that are not used.
- The cross_match name is now used (it defaults to `astromon_21`).
- Added the radial cut for grating observations. Until now it was the same as non-grating.

## Testing

- [x] Unit tests in mac
- [x] All the example code in [the docs](https://icxc.cfa.harvard.edu/aspect/tmp/jgonzalez/astromon/pr-17-docs). Before running the last example in the quick start, I did the following, which creates a test file:
   ```
   from astromon.tests import utils
   utils.create_h5('some_file.h5')
   ```
   **NOTE**: the observation code examples in the docs only run on HEAD and if you have access to arc5gl.
- [x] Functional test (running the production script). I did the following on HEAD and verified that the cross-matches are the same as before:
   ```
   python -m astromon.scripts.get_cat_obs_data --workdir work --obsid 26266 --db-file pr-17.h5
   ```
   and
   ```
   python -m astromon.scripts.get_cat_obs_data --workdir work --start 2022-01-04 --stop 2022-01-05 --db-file pr-17-2.h5
   ```
   and this one only writes xray_src and obs, the rest are empty
   ```
   python -m astromon.scripts.get_cat_obs_data --workdir work --obsid 26249 --db-file pr-17-3.h5
   ```

taldcroft reviewed (commented):
The 2nd, 3rd, and 4th examples in the Quick Start all give me exceptions. 

I vaguely recall that after I made an astromon-play-API notebook with my first use of the database, then you made a version using the new built-in API. Do you still have that, and can you update it to work with the API as it now exists in this PR?

javierggt commented on docs/quick.rst:
```diff
@@ -1,20 +1,79 @@
+--------------------
+
+A common usage of the astromon data products is to get the *standard* pre-computed cross-matches
+using :any:`db.get_cross_matches`. This function takes a `name` argument that specifies the
```
Only one of those examples failed for me, so it would be nice to know what error you got.

I fixed the one that was failing. It was because of inconsistency between the output from db.get_cross_matches and `cross_match.cross_match`.

Right when you were leaving your comments I added a new function. All the functionality you want should already be there.

taldcroft commented:
Not quite in this PR, but this code got my attention. It is making a view of a subset of a Table and then sorting within that view. Apparently that works, but it still strikes me as possibly sketchy. It is also a lot slower than it needs to be because of making lots of tables and sorting columns that don't need attention.
```
    matches = matches.group_by('obsid')
    for g in matches.groups:
        g.sort(['cat_order', 'dr'])
    result = table.vstack([g[0] for g in matches.groups])
```
This should be orders of magnitude faster and equivalent (famous last words):
```
mg = matches.group_by('obsid')
indices = mg.groups.indices
idxs = []
for i0, i1 in zip(indices[1:], indices[:-1]):
    idx = i0 + np.lexsort((mg['dr'][i0:i1], mg['cat_order'][i0:i1]))[0]
    idxs.append(idx)
results = mg[idxs]
```

javierggt commented:
> Not quite in this PR...

I will make an issue with this and follow-up after this PR.

taldcroft commented:
Digging further on the point about the cross match code above, two things:

1. It is actually broken and I confirmed that the sources returned don't adhere to the expected catalog priority ordering. For instance for obsid 1520 xid=6, there are matches for SDSS and 2MASS. Using `catalogs=['ICRS', 'Tycho2', '2MASS', 'SDSS', 'USNO-B1.0'],` should choose `2MASS`. My original `get_xcorr` gives the 2MASS match, but the result from this branch is the `SDSS` one.

3. It is extremely slow in a apples-to-apples comparison with my original code. The line below is intended to do no filtering at all (to match my original `get_xcorr(xray_src, cat_src, obs)` function). The code below takes about 3 minutes to complete while the original `get_xcorr` finishes in 4 seconds.
```
xcorr = cross_match.compute_cross_matches(
    catalogs=['ICRS', 'Tycho2', '2MASS', 'SDSS', 'USNO-B1.0'],
    dr=10000,
    snr=-10000,
    r_angle=10000,
    r_angle_grating=10000,
    near_neighbor_dist=-10000,
)
```

taldcroft commented on astromon/cross_match.py:
```diff
@@ -468,26 +699,87 @@ def _simple_cross_match(
+    matches = matches[matches['dr'] < dr]
 
     if len(matches) == 0:
         return []
```
Why are you turning the 0-length table that has all the right columns into a list? Better to have a consistent return type.

javierggt replied:
This was changed in a recent [commit](https://github.com/sot/astromon/pull/17/commits/c5fc7b9e0e30cecf1165e78283eb6e9e7113fd61). In order to be fully consistent, I had to make a few other changes.

taldcroft commented on astromon/cross_match.py:
```diff
@@ -468,26 +699,87 @@ def _simple_cross_match(
     if len(matches) == 0:
         return []
 
-    matches = matches.group_by('obsid')
```
This was the bug that was causing your version to return so many fewer sources. It was just returning one match per obsid.

javierggt replied:
yup, that's a bug.

Still, you got more than 10 times the number of matches, and I don't think this accounts for that difference.

taldcroft replied:
After applying  ff34529 I get exactly the same number of matches (68197) in an apples-to-apples comparison. In my original code `xcorr` is every rough match for the mission (no filtering at all).
```
xcorr = cross_match.compute_cross_matches(
    catalogs=['ICRS', 'Tycho2', '2MASS', 'SDSS', 'USNO-B1.0'],
     dr=10000,
    snr=-10000,
    r_angle=10000,
    r_angle_grating=10000,
    near_neighbor_dist=-10000,
)
```

taldcroft commented:
I fixed the discrepancies between my original API code and this version with ff34529.

taldcroft commented on astromon/db.py:
```diff
@@ -515,7 +519,7 @@ def set_formats(dat):
 
 
-def get_cross_matches(name='astromon_21', dbfile=None):
+def get_cross_matches(name='astromon_21', dbfile=None, **kwargs):
```
This function should not include `**kwargs` since as stated it is for getting a standard cross match. The problem is that the `**kwargs` filtering might be inconsistent with the `name` filter and give unexpected results (e.g. setting `snr` smaller than what is in `astromon_21`).

javierggt replied:
You should expect to get results in the standard sample **and** `snr >= 1.0` and that is not inconsistent.

taldcroft replied:
But the API or documentation does not say anywhere what is the standard cross match, so how as a user am I supposed to know what is there?  

Anyway, I'm saying that as a user it is totally *CONFUSING* that there are two functions which can be used to get filtered cross matches, but one of them starts with an implicit set of filter criteria that is not obvious.

The function documentation itself says that if you want your own set of filter parameters then use `compute_cross_matches`.
```
    If you want *non-standard* cross-match, with your own set of parameters, refer to the
    :any:`compute_cross_matches <cross_match.compute_cross_matches>` function.
```

So each function should do exactly one thing, and this is all that is needed:
- `get_cross_matches` gets a standard cross match by name
- `compute_cross_matches` computes a custom cross match

I will also say that for me as a user this result is completely unexpected:
```
In [11]: xc = get_cross_matches(snr=1.0)

In [12]: np.min(xc['snr'])
Out[12]: 3.0013847
```
As a developer who has spent hours reading the code I understand it but that does not change my initial reaction as a user.

taldcroft replied:
OK, I'm rescinding this comment, leave the API as it is.

taldcroft commented on astromon/db.py:
```diff
@@ -515,7 +519,7 @@ def set_formats(dat):
-def get_cross_matches(name='astromon_21', dbfile=None):
+def get_cross_matches(name='astromon_21', dbfile=None, **kwargs):
     """
     Get a standard cross-match of observations, x-ray sources and catalog counterparts in `dbfile`.
```
I could not find the actual filter values of the standard cross-matches anywhere in the documentation. This is important information.

taldcroft commented on astromon/db.py:
```diff
@@ -540,21 +544,41 @@ def get_cross_matches(name='astromon_21', dbfile=None):
 
     Returns
     -------
     :any:`tables.File <tables.file.File>`
```
This function returns a `Table` not a `tables.File`.

taldcroft commented on astromon/db.py:
```diff
@@ -195,27 +195,31 @@ def get_table(table_name, dbfile=None):
-                res = table.Table((res))
-                res.convert_bytestring_to_unicode()
-                return res
+                result = table.Table((res))
```
Why the extra parentheses?

javierggt replied:
no idea!

taldcroft commented on astromon/db.py:
```diff
@@ -195,27 +195,31 @@ def get_table(table_name, dbfile=None):
             except tables.NoSuchNodeError:
                 if table_name in DTYPES:
                     # logger.warning(f'{table_name} is not in file. Returning empty table.')
                     warnings.warn(f'{table_name} is not in file. Returning empty table.')
```
This should be an exception handled by the code below. If the h5 file really does not have an expected table then something is badly wrong.

javierggt replied:
I agree that this is not ideal. I will add a method to explicitly create tables with the right dtype and remove this.

javierggt replied:
I did this in a recent [commit](https://github.com/sot/astromon/pull/17/commits/0cbe0c69271dbe27949f72e7b70f50741f6b7f35). I had to change a few other things, since this changes the default behavior.

taldcroft commented:
If I'm understanding correctly this code worries me:
```
        dtype = DTYPES[table_name]
        names = [n for n in dtype.names if n in data.dtype.names]
        if len(names) == 0:
            raise Exception('Input data has no columns in common with table in DB')
        b = np.zeros(len(data), dtype=dtype)
        b[names] = data[names].as_array().astype(dtype[names])
        data = b
```
It basically says that the calling function can supply a `data` table that does not have the expected columns and the missing columns will silently be filled with zeros. I think that finding the expected columns should be required. The following code does this (I think):
```
dtype = DTYPES[table_name]
data = data[dtype.names].as_array().astype(dtype)
```

taldcroft commented:
@javierggt - this PR is basically fine at this point except for the minor comments I made that are all easily fixed. If you just take care of those then we can merge this one.

javierggt commented:
I made changes that I believe address @taldcroft's comments.

I also made a couple new commits:
- removed an unused script
- changed the argument orders in `db.save` to be consistent
- updated test data

taldcroft commented:
I pushed a commit so the narrative docs include the information that future-me will need.

taldcroft reviewed (commented):
Just one minor comment which might be ignorable.

javierggt commented:
I committed a fix for a failure I got, which I believe traces back to changing the return values of empty tables. I didn't spend much time checking why, but the bottomline is I add some extra columns to empty tables so the result's type is consistent with the case of non-empty tables, and downstream code can handle it.

I ran the tests above and check that the output is consistent with expectations.

taldcroft reviewed (approved):
Looks good thanks for all the effort!

javierggt commented:
I fixed two typos and will squash in a moment.

taldcroft commented:
:tada:
